docs/factories.rst

```diff
@@ -0,0 +1,353 @@
+
+The ``for_`` used defaults again to the class.
+
+The imperative form, instead of a decorator:
+
+.. code-block:: python
+
+    registry.register_factory(LoginService)
+
+Simple DI By-Type
+=================
+
+That's nice, but it's still a bit of boilerplate.
+The new factory support handles dependency injection in a variety of useful ways.
+Let's show how the class can simply say "Give me the Session":
+
+.. code-block:: python
+
+    @wired_factory()
+    class LoginService:
+        dbsession: Session
+        def __init__(self, dbsession: Session):
+            self.dbsession = dbsession
+
+In this case, the constructor's (one) argument was a type that is registered as a service.
+``@wired_factory`` sniffed at the argument type, did a ``container.get()``, and used that as the constructor argument.
+
+Let's see a dataclass flavor of that, to reduce the boilerplate even more:
+
+.. code-block:: python
+
+    @wired_factory()
+    @dataclass
+    class LoginService:
+        dbsession: Session
+
+Again, this works with either the ``@wired_factory`` decorator or the imperative ``register_factory``.
+Injection also works with functions:
+
+.. code-block:: python
+
+    @factory
+    def login_factory(dbsession: Session) -> LoginService:
+        # Potentially do some work here
+        return LoginService(dbsession=dbsession)
+
+Maybe "Potentially do some work here" means access to the container.
+If so, just ask for it as well:
+
+.. code-block:: python
+
+    @wired_factory()
+    def login_factory(container: ServiceContainer, dbsession: Session) -> LoginService:
+        # Potentially do some work here
+        return LoginService(dbsession=dbsession)
+
+Argument order doesn't matter:
+
+.. code-block:: python
+
+    @wired_factory()
+    def login_factory(dbsession: Session, container: ServiceContainer) -> LoginService:
+        # Potentially do some work here
+        return LoginService(dbsession=dbsession)
+
+A DSL for Injection: Type Changing
+==================================
+
+*Note: This might be where the default injector parts ways with a ``@wired_dataclass`` or ``__wired_factory__`` add-on.*
+
+The container's context is a useful thing to have in a service.
+The DI story makes this easy, using the ``Context`` marker which flags the injector to get ``container.context``:
+
+.. code-block:: python
+
+    @wired_factory()
+    @dataclass
+    class Greeting:
+        customer: Context
+
+        def __call__(self) -> str:
+            return f'Hello {self.customer.name}'
+
+That's great, but sometimes we know the context is a ``Customer`` and we want to use that in the type hint.
+It would be nice if we could tell *Python* one type hint, but use another type for injection.
+
+Let's use PEP 593 "Flexible Annotations".
+``wired`` would add:
+
+.. code-block:: python
+
+    Injector = object()
+    LT = TypeVar("local_type")
+    RT = TypeVar("registered_type")
+    Injected = Annotated[LT, RT, Injector]
```

Comment:
This is getting pretty wild for me as someone who is barely even using typing. It's interesting though!